airo-robots/airo_robots/manipulators/position_manipulator.py

```diff
@@ -205,6 +208,67 @@ def is_tcp_pose_reachable(self, tcp_pose: HomogeneousMatrixType) -> bool:
 
+    def execute_trajectory(self, joint_trajectory: SingleArmTrajectory) -> None:
+        """Execute a joint trajectory. This function will interpolate the trajectory and send the commands to the robot.
+
+        This function is implemented according to the notes of https://github.com/airo-ugent/airo-mono/issues/150.
+        Please refer to this issue for design decisions.
+
+        Args:
+            joint_trajectory: the joint trajectory to execute."""
+        self._assert_joint_trajectory_is_executable(joint_trajectory)
+
+        period = 0.005  # Time per servo, approximately. This may be slightly changed because of rounding errors.
+        # The period determines the times at which we sample the trajectory that was time-parameterized.
+        duration = (joint_trajectory.times[-1] - joint_trajectory.times[0]).item()
+
+        n_servos = int(np.ceil(duration / period))
+        period_adjusted = duration / n_servos  # can be slightly different from period due to rounding
+
+        start_time_ns = time.time_ns()
+        for servo_index in range(n_servos):
+            current_time_ns = time.time_ns()
+            t_ns = current_time_ns - start_time_ns
+            t = t_ns / 1e9
+            if t > duration:
+                logger.warning(
+                    f"Time exceeded trajectory duration at servo index {servo_index} / {n_servos}. This means we are lagging, but we should have reached the final configuration. Stopping trajectory execution."
+                )
+                break
+
+            # Find the two joint configurations that are closest to time t.
+            i0 = np.searchsorted(joint_trajectory.times, t, side="left") - 1  # - 1: i0 is always >= 1 otherwise.
+            i1 = i0 + 1
+
+            if i1 == len(joint_trajectory.times):
+                break
+
+            # Interpolate between the two joint configurations.
+            q_interp = lerp_positions(i0, i1, joint_trajectory.path.positions, joint_trajectory.times, t)
+            self.servo_to_joint_configuration(q_interp, period_adjusted)
+            # We do not wait for the servo to finish, because we want to sample the trajectory at a fixed rate and avoid lagging.
+
+            if joint_trajectory.gripper_path is not None:
+                if self.gripper is None:
+                    raise ValueError("Gripper trajectory provided, but no gripper is attached to the manipulator.")
+
+                gripper_pos_interp = lerp_positions(
+                    i0, i1, joint_trajectory.gripper_path.positions, joint_trajectory.times, t
+                ).item()
+                self.gripper.move(gripper_pos_interp)
+
+            # time.sleep(
+            #     period_adjusted
+            # )
+
+        # This avoids the abrupt stop and "thunk" sounds at the end of paths that end with non-zero velocity
+        # TODO: This is not an attribute of PositionManipulator, but is specific to URrtde. How do we implement this correctly?
```

Comment:
could add this function to the interface? But do we want to have this as default behavior? shouldn't the user ensure the trajectory is well-behaved?

Reply:
If we do not add the function to the interface (and use it here) I believe a warning or at the very least a comment in the method's docstring would be appropriate, to avoid mistakes when new users execute trajectories that end with non-zero velocities.